app/controllers/spina/registers_controller.rb

```diff
@@ -54,36 +56,77 @@ def history
       @register = Spina::Register.find_by_slug!(params[:id])
       @register_data = @@registers_client.get_register(@register.name.parameterize, @register.register_phase)
 
-      fields = @register_data.get_field_definitions.map{ |field| field[:item]['field'] }
+      fields = @register_data.get_field_definitions.map { |field| field[:item]['field'] }
       all_records = @register_data.get_records_with_history
       entries_reversed = @register_data.get_entries.reverse
 
-      entries_mapped_with_items = entries_reversed.map { |entry|
+      entries_mapped_with_items = entries_reversed.map do |entry|
         records_for_key = all_records[entry[:key]]
-        current_record = records_for_key.detect{ |record| record[:entry_number] == entry[:entry_number] }
+        current_record = records_for_key.detect { |record| record[:entry_number] == entry[:entry_number] }
         previous_record_index = records_for_key.find_index(current_record) - 1
 
         if previous_record_index < 0
           changed_fields = fields
         else
           previous_record = records_for_key[previous_record_index]
-          changed_fields = fields.select { |f| current_record[:item][f] != previous_record[:item][f] }
+          changed_fields = fields.reject { |f| current_record[:item][f] == previous_record[:item][f] }
         end
 
         { current_record: current_record, previous_record: previous_record, updated_fields: changed_fields, key: entry[:key] }
-      }
+      end
 
-      @entries_with_items = paginate(entries_mapped_with_items)
+      if params[:q].present?
+        filtered = filter(entries_mapped_with_items, params[:q])
+        @entries_with_items = paginate(filtered)
+      else
+        @entries_with_items = paginate(entries_mapped_with_items)
+      end
     end
 
     private
 
-    def initialize_client
-      @@registers_client ||= RegistersClient::RegistersClientManager.new({ cache_duration: 600 })
+    def filter(entries, query)
+      entries.select do |entry|
+        entry[:key].downcase.include?(query.downcase) ||
+          contained?(entry[:previous_record], query) ||
+          contained?(entry[:current_record], query)
```

Comment:
This searches all values in the before/after items for the record, not only those that have changed. The effect is that some results in the search do not actually contain the thing you have searched for in the visual "diff" (though they would if the whole of the items were displayed). For example on the updates page for the country page, search for "republic" and you will see that there is an entry for `GM` that does not actually visually display the query string.

Is this intentional? Or maybe it's part of the use case we hadn't considered yet - if so I will speak to design.

Reply:
I've also noticed that this search searched on both field-name and field-value. Is it intentional that it also searches field-name? I can see how this might make sense for the updates page (if someone is searching for entries where a particular field-name has been updated) but this does not make sense for the records page. Currently, in production (https://registers-trial.service.gov.uk/registers/country) if you search for "official" you get no results, but in this branch you get all results because all rows have a field-name called "official-name".

This may mean that we have to split the logic slightly for records search and updates search (sorry, my fault for not thinking ahead). However, I don't think searching the updates page by field-name makes sense if we are also searching the whole item (as opposed to only those that have changed), which would involve changing the logic I mentioned in the comment above.